BloomFilter/Program.cs

```diff
@@ -0,0 +1,346 @@
+            }, ct);
+
+            while (true)
+            {
+                var keyPress = Console.ReadKey();
+                if (keyPress.KeyChar == 'q')
+                {
+                    break;
+                }
+            }
+
+            logger.DebugFormat("Exiting...");
+
+            if (node != null)
+            {
+                node.DisconnectAsync();
+            }
+
+            tokenSource.Cancel();
+        }
+
+        /// <summary>
+        /// Attempts to load the persisted blockchain headers from persistent storage. This is 
+        /// a lot quicker than loading from a peer node.
+        /// </summary>
+        private async static Task LoadChain(ConcurrentChain chain, CancellationToken ct)
+        {
+            logger.DebugFormat("Commencing blockchain headers load from disk...");
+
+            ct.ThrowIfCancellationRequested();
+
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+
+            if (File.Exists(_chainFile))
+            {
+                await Task.Run(() => { chain.Load(File.ReadAllBytes(_chainFile)); });
+            }
+
+            sw.Stop();
+
+            logger.DebugFormat("Block headers load from disk, chain height {0} in {1}s.", chain.Height, sw.Elapsed.Seconds);
+        }
+
+        /// <summary>
+        /// Peridically persists any updated blockchain headers to disk.
+        /// </summary>
+        private static void PersistChain(ConcurrentChain chain, CancellationToken ct)
+        {
+            logger.DebugFormat("Starting persist blockchain task.");
+
+            ct.ThrowIfCancellationRequested();
+
+            int chainHeight = (chain != null) ? chain.Height : 0;
+
+            while (ct.IsCancellationRequested == false)
+            {
+                if (chain.Height != chainHeight)
+                {
+                    using (var fs = File.Open(_chainFile, FileMode.Create))
+                    {
+                        chain.WriteTo(fs);
+                    }
+
+                    logger.DebugFormat("Chain height increased to {0} ({1})", chain.Height, DateTime.Now.ToString("HH:mm:ss"));
+                    chainHeight = chain.Height;
+                }
+
+                Task.Delay(5000, ct).Wait();
+            }
+        }
+
+        /// <summary>
+        /// Attempts to connect to a full BitCoin node and request any missing block headers.
+        /// </summary>
+        private static async Task<Node> ConnectNodeAndSyncHeaders(ConcurrentChain chain, CancellationToken ct)
+        {
+            logger.DebugFormat("Connecting to full node...");
+
+            ct.ThrowIfCancellationRequested();
+
+            ManualResetEventSlim headersSyncedSignal = new ManualResetEventSlim();
+            var parameters = new NodeConnectionParameters();
+            parameters.IsRelay = false;
+
+            NodeRequirement req = new NodeRequirement();
+            req.RequiredServices = NodeServices.NODE_BLOOM;
+            req.SupportSPV = true;
```

Comment:
Can you declare `req` next to where you use it?
Actually this code is optional, since your `bitcoind` supports bloomfilter, this code only make sure it does during the handshake (ie, if you disable bloomfilter with `-peerbloomfilters=0`, then the handshake will fail)